packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1933,12 +2010,29 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
 
   @override
   void didChangeNext(Route<dynamic>? nextRoute) {
+    if (nextRoute is ModalRoute<T> && nextRoute.delegatedTransition == null && nextRoute.subtreeContext == null && this.subtreeContext != null) {
+      nextRoute.setDelegatedTransition(this.subtreeContext!);
+    }
+    if (nextRoute is ModalRoute<T> && canTransitionTo(nextRoute) && nextRoute.delegatedTransition.runtimeType != this.delegatedTransition.runtimeType) {
+      if (receivedTransition.runtimeType != nextRoute.delegatedTransition.runtimeType) {
+        receivedTransition = nextRoute.delegatedTransition;
+      }
```

Comment:
Is this handling both giving this route's delegatedTransition to the nextRoute and receiving a delegatedTransition from the nextRoute? Why not handle the "giving" part in didChangePrevious?

Reply:
`nextRoute.setDelegatedTransition` is not passing anything to the next route. There are some cases where didChangeNext is triggered before the next page builds, and the next page needs to know it's build context to get the delegated transition. For example, Material pages get their transitions from the theme. So `setDelegatedTransition` gets the next page to set a delegated transition based on the previous page's context. Maybe `setDelegatedTransition` could have a better name. I went looking through the route's lifecycle as it's added to the stack and couldn't find a better place to do this.

Reply:
Note from offline: add comment explaining.